File: solutions/1915. Number of Wonderful Substrings/1915.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  long long wonderfulSubstrings(string_view word) {
    long long ans = 0;
    int prefix = 0;
    vector<int> count(1024);
    count[0] = 1;

    for (const char c : word) {
      prefix ^= 1 << c - 'a';
      ans += count[prefix];
      for (int i = 0; i < 10; ++i)
        ans += count[prefix ^ 1 << i];
      ++count[prefix];
    }

    return ans;
  }
};

File: solutions/1915. Number of Wonderful Substrings/1915.rs
// Time complexity: O(n)
// Space complexity: O(1)
impl Solution {
    pub fn wonderful_substrings(word: String) -> i64 {
        let mut ans = 0;
        let mut prefix = 0;
        let mut count: [i64; 1024] = [0; 1024];
        count[0] = 1;

        for c in word.chars() {
            prefix ^= 1 << (c as u8 - b'a');
            ans += count[prefix];
            for i in 0..10 {
                ans += count[prefix ^ (1 << i)];
            }
            count[prefix] += 1;
        }

        ans
    }
}
